django/cantusdb_project/main_app/permissions.py

```diff
@@ -0,0 +1,130 @@
+from django.db.models import Q
+
+
+def user_can_edit_chants_in_source(user, source) -> bool:
+    """
+    Checks if the user can edit Chants in a given Source.
+    Used in ChantDetail, ChantList, ChantCreate, ChantDelete, ChantEdit,
+    ChantEditSyllabification, and SourceDetail views.
+    """
+    if user.is_anonymous or (source is None):
+        return False
+
+    source_id = source.id
+    user_is_assigned_to_source: bool = user.sources_user_can_edit.filter(
+        id=source_id
+    ).exists()
+
+    user_is_project_manager: bool = user.groups.filter(name="project manager").exists()
+    user_is_editor: bool = user.groups.filter(name="editor").exists()
+    user_is_contributor: bool = user.groups.filter(name="contributor").exists()
+
+    return (
+        (user_is_project_manager)
+        or (user_is_editor and user_is_assigned_to_source)
+        or (user_is_editor and source.created_by == user)
+        or (user_is_contributor and user_is_assigned_to_source)
+        or (user_is_contributor and source.created_by == user)
+    )
+
+
+def user_is_proofreader(user) -> bool:
```

Comment:
Do we use "proofreader" anywhere else in the code base? I thought this was only used in OldCantus, and that OC "proofreaders" are now NC "editors".

A better name for this function might be `user_can_access_proofreading_pages` or something similar

Reply:
Or maybe it would make sense to get rid of this function, and use `user_can_proofread_chants_in_source` in all situations? Not sure if that works.

Reply:
This is possible now, needed to do a bit of reworking in `UserSourceListView` to add a list of tuples (source, permission for source) to the pagination object.